auth/service.cc

```diff
@@ -423,6 +426,230 @@ future<bool> service::exists(const resource& r) const {
+    //     but case sensitivity can be forced by using double-quotes ("myTable" is different from mytable).
+    //
+    // That's why we wrap identifiers with quotation marks below.
+
+    if (!maybe_ks) {
+        return seastar::format("GRANT {} ON ALL FUNCTIONS TO {};", permission, formatted_role);
+    }
+    const auto ks = cql3::util::maybe_quote(*maybe_ks);
+
+    if (!maybe_fun_sig && !maybe_fun_name) {
+        return seastar::format("GRANT {} ON ALL FUNCTIONS IN KEYSPACE {} TO {};", permission, ks, formatted_role);
+    }
+
+    if (maybe_fun_name) {
+        SCYLLA_ASSERT(maybe_fun_args);
+
+        const auto fun_name = cql3::util::maybe_quote(*maybe_fun_name);
+        const auto fun_args_range = *maybe_fun_args | std::views::transform([] (const auto& fun_arg) {
+            return cql3::util::maybe_quote(fun_arg);
+        });
+
+        return seastar::format("GRANT {} ON FUNCTION {}.{}({}) TO {};",
+                permission, ks, fun_name, fmt::join(fun_args_range, ", "), formatted_role);
+    }
+
+    SCYLLA_ASSERT(maybe_fun_sig);
+
+    auto [fun_name, fun_args] = decode_signature(*maybe_fun_sig);
+    fun_name = cql3::util::maybe_quote(fun_name);
+
+    // We don't call `cql3::util::maybe_quote` later because `cql3_type_name_without_frozen` already guarantees
+    // that the type will be wrapped within double quotation marks if it's necessary.
+    auto parsed_fun_args = fun_args | std::views::transform([] (const data_type& dt) {
+        return dt->without_reversed().cql3_type_name_without_frozen();
+    });
+
+    return seastar::format("GRANT {} ON FUNCTION {}.{}({}) TO {};",
+            permission, ks, fun_name, fmt::join(parsed_fun_args, ", "), formatted_role);
+}
+
+// The function doesn't assume anything about `role`.
+static sstring describe_resource_kind(const permission& perm, const resource& r, std::string_view role) {
+    switch (r.kind()) {
+        case resource_kind::data:
+            return describe_data_resource(perm, r, role);
+        case resource_kind::role:
+            return describe_role_resource(perm, r, role);
+        case resource_kind::service_level:
+            return describe_service_level_resource(perm, r, role);
+        case resource_kind::functions:
+            return describe_udf_resource(perm, r, role);
+    }
+}
+
+future<std::vector<cql3::description>> service::describe_permissions() const {
+    std::vector<cql3::description> result{};
+
+    const auto permission_list = co_await std::invoke([&] -> future<std::vector<permission_details>> {
+        try {
+            co_return co_await _authorizer->list_all();
+        } catch (const unsupported_authorization_operation&) {
+            // If Scylla uses AllowAllAuthorizer, permissions do not exist and the corresponding authorizer
+            // will throw an exceptoion when trying to access them.
+            co_return std::vector<permission_details>{};
+        }
```

Comment:
It's a bad practice to use exceptions for control flow. They should only be thrown in truly unexpected scenarios. Throwing an exception also increments a metric which is exposed to users, this will result in very confusing behavior - when you DESCRIBE and have AllowAllAuthorizer, there will be an unexplained spike of exceptions, and you will have to deal with reports from confused users, customers or even other developers who suspect that there is a bug somewhere.

Can we do it somehow without involving exceptions? cc: @nuivall

Reply:
In theory, we could add a function indicating whether a given authoriser supports permissions and replace the `try-catch` with an `if`. The question is if it's a good design decision

Reply:
I see this pattern used elsewhere, so it's a wider problem. I think it would be good to get rid of those throws eventually; however, right now it's just a single throw for each DESCRIBE - and it only happens if authentication is not enabled - so I think it's not a huge problem and we can postpone it.

Reply:
technically it's not an exception  (not 100% sure but I think there is no catch on this) but rather unreachable code marker, if we have other way we could use it (abort? or SCYLLA_ABORT - although I am not confident enough to kill the whole process)

Reply:
@nuivall if you grep through the code you will notice that we `catch (const unsupported_authorization_operation&)` in several places, and then ignore or translate it to a different type of exception - not sure if from the `list_all` method specifically, but some other methods in the authorizer interface have similar semantics (i.e. throw this exception if the operation is not supported).

Reply:
ok, yes, it's weird. sometimes they just catch and ignore it